A67-xds-dynamic-parameters.md

```diff
@@ -0,0 +1,246 @@
+>      }
+     }
+   },
+   "node": {
+     "id": "projects/123456789123/networks/default/nodes/4d5f8bd2-3e91-4940-855a-fdabca4c2f70"
+   }
+ }
+```
+
+To support non-xdstp use cases, we can add a map at the root level too:
+
+```text
+ {
+   "xds_servers": [
+     {
+       "server_uri": "xds.example.com:443",
+       "channel_creds": [ { "type": "google_default" } ]
+       ],
+       "server_features": [
+         "xds_v3"
+       ]
+     }
+   ],
+   "node": {
+     "id": "projects/123456789123/networks/default/nodes/4d5f8bd2-3e91-4940-855a-fdabca4c2f70"
+   },
+>  "dynamic_parameters": {
+>    "env": "prod"
+>  }
+ }
+```
+
+The `XdsClient` is currently structured as follows in all languages:
+
+-   There is a bootstrap config known by the `XdsClient` object. The bootstrap
+    config contains settings for each authority.
+-   We have a shared pool of xDS channel objects, so if multiple authorities use
+    the same xDS server, they will share the same xDS channel object.
+-   The code that constructs the ADS request already knows the set of resource
+    names to send, and that list may include resources from multiple authorities
+    that are sharing the same xDS server.
+
+The changes this design proposes are:
+
+-   The bootstrap config will contain the dynamic parameters to use for each
+    authority.
+-   The code to construct the ADS request will be changed such that for each
+    authority for which there are resource names in the request, it looks up the
+    dynamic parameters for that authority in the bootstrap config. If the
+    dynamic parameters are not empty, that will trigger using the
+    `resource_locators` field instead of the `resource_names` field in the
+    request
+-   The code to parse the ADS response will be changed to look in the
+    `resource_name` field and use this instead of the name field if it exists.
+    As this `XdsClient` is only a leaf client, and not a caching proxy, we can
+    ignore the dynamic_parameter_constraints field of the ResourceName message.
+
+The updated code to construct the ADS request would look something like this
+(pseudo code):
+
+```text
+// Example authority_resource_names:
+[
+  // Resources for authority "foo.example.com"
+  { resource_names: ["bar", "baz"], dynamic_parameters: {"env": "prod"} },
+
+  // Resources for authority "zzz.example.com"
+  { resource_names: ["yyy"], dynamic_parameters: {"env": "prod", "version": "v2"} },
+]
+
+function ConstructADSRequest(authority_resource_names)
+  ...
+  foreach resource_names, dynamic_parameters in authority_resource_names
+    if dynamic_parameters is empty
+      request.resource_names = resource_names
+    else
+      foreach name in resource_names
+        request.resource_locators.add(name, dynamic_parameters)
+  ...
+```
+
+### Temporary environment variable protection
+
+During initial development, the GRPC_EXPERIMENTAL_XDS_DYNAMIC_PARAMETERS
+environment variable will control whether ADS requests will use the
```

Comment:
The overall goal of the env var is to ensure that until the feature is tested and considered stable, we don't want to enable any new functionality that can be triggered via something out of gRPC's control.  And we need to be very precise here about what parts of the code are guarded by the env var, because we've had bugs in the past where we failed to guard things, and we really don't want to repeat that mistake.

Note that the goal here is *not* to catch problems where the bootstrap config is updated to use dynamic parameters but the xDS server doesn't support them.  I would not actually expect that case to happen in practice, because whoever maintains the bootstrap file should not configure it to use a feature with a server that doesn't support that feature.  And if they did do that, there's really nothing we can do to avoid it.

Also, note that this document is for OSS gRPC, but the bootstrap generator you're linking to here is TD-specific.  It's fine to cite that as an example, but I think we should describe the problem more generally: it happens in cases where the bootstrap file is generated by a separate tool with an independent release process.

So, putting all of this together, I suggest rewording this paragraph as follows:

"""
During development, the dynamic parameter functionality will be guarded by the GRPC_EXPERIMENTAL_XDS_DYNAMIC_PARAMETERS env var.  Specifically, this env var will control two things:
- Whether we read the new `dynamic_parameters` fields in the bootstrap config (and therefore whether we populate the `resource_locators` field in ADS requests).  This guards against cases where the bootstrap file is generated by a separate tool with a release process independent of that of the application (e.g., the [Traffic Director bootstrap generator](https://github.com/GoogleCloudPlatform/traffic-director-grpc-bootstrap)), such that a newer bootstrap config may be used with an older version of gRPC.
- Whether we look at the new `resource_names` field in ADS responses.  This guards against xDS servers that may incorrectly start sending new fields to clients that are not yet ready to understand them.

The env var protection will be removed once this feature passes interop tests.
"""

Reply:
Thanks for the rewording suggestion. Applied.